File: pages/login_page.py
from playwright.sync_api import Locator, Page

from pages.base_page import BasePage


class LoginPage(BasePage):
    PATH = "/"

    def __init__(self, page: Page, base_url: str) -> None:
        super().__init__(page, base_url)
        self._username: Locator = page.locator("#user-name")
        self._password: Locator = page.locator("#password")
        self._submit: Locator = page.locator("#login-button")

    def open(self) -> None:
        self.goto(self.PATH)

    def login(self, username: str, password: str) -> None:
        self._username.fill(username)
        self._password.fill(password)
        self._submit.click()

    @property
    def error_message(self) -> Locator:
        return self.page.locator("[data-test='error']")

File: java/selenium/src/test/java/com/tnd/selenium/pages/LoginPage.java
package com.tnd.selenium.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * Page object for SauceDemo login screen (Selenium Java).
 *
 * <p>Purpose: encapsulate authentication interactions reused across flow tests.
 */
public class LoginPage extends BasePage {
    private static final By USERNAME = By.id("user-name");
    private static final By PASSWORD = By.id("password");
    private static final By LOGIN = By.id("login-button");

    /**
     * Creates login page object.
     */
    public LoginPage(WebDriver driver, String baseUrl) {
        super(driver, baseUrl);
    }

    /**
     * Opens login route.
     */
    public void open() {
        goTo("/");
    }

    /**
     * Submits credentials for authentication flow.
     */
    public void login(String username, String password) {
        fill(USERNAME, username);
        fill(PASSWORD, password);
        click(LOGIN);
    }
}
